text/0000-float-semantics.md

```diff
@@ -0,0 +1,290 @@
+- Feature Name: `float_semantics`
+- Start Date: 2023-10-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a specification for how our floating point operations are expected to behave.
+The current implementation in rustc already matches the specification, so after accepting the RFC no compiler changes are required.
+(However we might be able to stabilize some `const fn` features, see below for details.)
+
+Rust's floating point operations follow IEEE 754-2008 -- with some caveats around operations producing NaNs: IEEE makes almost no guarantees about the sign and payload bits of the NaN; however, actual hardware does not pick those bits completely arbitrarily, and Rust will expose some of those hardware-provided guarantees to programmers.
+On the flip side, NaN generation is non-deterministic: running the same operation on the same inputs several times can produce different results.
+And there is a caveat: while IEEE specifies that float operations can never output a signaling NaN, Rust float operations *can* produce signaling NaNs, *but only if* an input is signaling.
+That means the only way to ever see a signaling NaN in a program is to create one with `from_bits` (or equivalent unsafe operations).
+
+Floating-point operations at compile-time follow the same specification. In particular, since operations are non-deterministic, the same operation can lead to different bit-patterns when executed at compile-time (in a `const` context) vs at run-time.
```

Comment:
```suggestion
Floating-point operations at compile-time follow the same specification. In particular, since operations involving NaN bit patterns are non-deterministic, the same operation can lead to different NaN bit patterns when executed at compile-time (in a `const` context) vs at run-time.
```

To make it more obvious this RFC is only documenting non-determinism for NaNs, not for floating point in general.